src/stream/src/executor/arrangement_backfill.rs

```diff
@@ -0,0 +1,679 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::ops::Bound;
+use std::pin::pin;
+use std::sync::Arc;
+
+use await_tree::InstrumentAwait;
+use either::Either;
+use futures::stream::select_with_strategy;
+use futures::{pin_mut, stream, StreamExt, TryStreamExt};
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::array::stream_record::Record;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::buffer::BitmapBuilder;
+use risingwave_common::catalog::Schema;
+use risingwave_common::hash::VnodeBitmapExt;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_common::types::Datum;
+use risingwave_common::util::epoch::EpochPair;
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_common::util::sort_util::{cmp_datum, OrderType};
+use risingwave_storage::table::collect_data_chunk;
+use risingwave_storage::StateStore;
+
+use super::error::StreamExecutorError;
+use super::{expect_first_barrier, BoxedExecutor, Executor, ExecutorInfo, Message, PkIndicesRef};
+use crate::common::table::iter_utils::merge_sort;
+use crate::common::table::state_table::StateTable;
+use crate::executor::monitor::StreamingMetrics;
+use crate::executor::{Barrier, PkIndices, StreamExecutorResult, Watermark};
+use crate::task::{ActorId, CreateMviewProgress};
+
+pub struct ArrangementBackfillExecutor<S: StateStore> {
+    /// Upstream table
+    upstream_table: StateTable<S>,
```

Comment:
This state table looks like a normal one instead of the replicated state table.

Reply:
I'm thinking that we can just add a flag to the state table, which means the type would be the same still.

Similar to how we handle replicated local hummock state store.

Reply:
Another option is to add a generic const like `StateTable<const IS_REPLICATED: bool=false, S>` to the StateTable, since the flag is determined statically at compile time.

Reply:
Fixed